### PR TITLE
[Merged by Bors] - feat: more lemmas about `List.dedup` and other lattice operations

### PR DESCRIPTION
**Pull request by eric-wieser**
Also the trivial `Multiset` wrappers.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Dedup.lean [hunk 1]
```diff
@@ -132,6 +133,41 @@ theorem dedup_append (l₁ l₂ : List α) : dedup (l₁ ++ l₂) = l₁ ∪ ded
   · rw [dedup_cons_of_not_mem' h, insert_of_not_mem h]
 #align list.dedup_append List.dedup_append
 
+theorem dedup_map_of_injective [DecidableEq β] {f : α → β} (hf : Function.Injective f)
+    (xs : List α) :
+    (xs.map f).dedup = xs.dedup.map f := by
```
Comment (YaelDillies) [inline Mathlib/Data/List/Dedup.lean:138]: @Timeroot had this lemma somewhere. Did it ever land? What is the PR?

File: Mathlib/Data/List/Dedup.lean [hunk 2]
```diff
@@ -132,6 +133,41 @@ theorem dedup_append (l₁ l₂ : List α) : dedup (l₁ ++ l₂) = l₁ ∪ ded
...
+  induction xs with
+  | nil => simp
+  | cons x xs ih =>
+    rw [map_cons]
+    by_cases h : x ∈ xs
+    · rw [dedup_cons_of_mem h, dedup_cons_of_mem (mem_map_of_mem f h), ih]
+    · rw [dedup_cons_of_not_mem h, dedup_cons_of_not_mem <| (mem_map_of_injective hf).not.mpr h, ih,
+        map_cons]
+
+theorem Subset.dedup_append {xs ys : List α} (h : xs ⊆ ys) :
```
Comment (YaelDillies) [inline Mathlib/Data/List/Dedup.lean:148]:
```suggestion
theorem Subset.dedup_append_left {xs ys : List α} (h : xs ⊆ ys) :
```
and can we have the right version stated using `List.Perm`?

File: Mathlib/Data/Multiset/FinsetOps.lean [hunk 3]
```diff
@@ -280,6 +289,12 @@ theorem ndinter_eq_zero_iff_disjoint {s t : Multiset α} : ndinter s t = 0 ↔ D
   rw [← subset_zero]; simp [subset_iff, Disjoint]
 #align multiset.ndinter_eq_zero_iff_disjoint Multiset.ndinter_eq_zero_iff_disjoint
 
+alias ⟨_, Disjoint.ndinter_eq_zero⟩ := ndinter_eq_zero_iff_disjoint
+
+theorem Subset.ndinter_eq_left {s t : Multiset α} (h : s ⊆ t) : s.ndinter t = s := by
```
Comment (YaelDillies) [inline Mathlib/Data/Multiset/FinsetOps.lean:294]: Same, can we have the right verSion?

File: Mathlib/Data/List/Dedup.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Data/List/Dedup.lean:148]: Hmm maybe, but the fact it's missing compared to the other direction is quite upsetting and might cause confusion in the future. At least add a comment explaining why the other lemma doesn't exist

Comment (eric-wieser) [reply Mathlib/Data/List/Dedup.lean]: I added the lemma, but in the multiset file where the proof was trivial.

File: Mathlib/Data/Multiset/FinsetOps.lean (hunk 3 above)
Comment (eric-wieser) [reply Mathlib/Data/Multiset/FinsetOps.lean:294]: I don't think the right version is meaningfully true? What statement do you want?

Comment (YaelDillies) [reply Mathlib/Data/Multiset/FinsetOps.lean:294]: Ah indeed. But then this lemma is something like `Multiset.filter_true_of_mem h` except that `filter_true_of_mem` doesn't exist

File: Mathlib/Data/Multiset/Dedup.lean [hunk 4]
```diff
@@ -132,6 +137,25 @@ theorem Nodup.le_dedup_iff_le {s t : Multiset α} (hno : s.Nodup) : s ≤ t.dedu
...
+theorem Subset.dedup_add_left {s t : Multiset α} (h : t ⊆ s) :
+    dedup (s + t) = dedup s := by
+  rw [add_comm, Subset.dedup_add_right h]
+
+theorem Disjoint.dedup_add {s t : Multiset α} (h : Disjoint s t) :
+    dedup (s + t) = dedup s + dedup t := by
+  induction s, t using Quot.induction_on₂
+  exact congr_arg ((↑) : List α → Multiset α) <| List.Disjoint.dedup_append h
+
+/-- Note that the tronger `List.Subset.dedup_append_right` is proved earlier. -/
```
Comment (YaelDillies) [inline Mathlib/Data/Multiset/Dedup.lean:154]:
```suggestion
/-- Note that the stronger `List.Subset.dedup_append_right` is proved earlier. -/
```

File: Mathlib/Data/Multiset/FinsetOps.lean (hunk 3 above)
Comment (eric-wieser) [reply Mathlib/Data/Multiset/FinsetOps.lean:294]: It exists for `List` though; I golfed the proof there

File: Mathlib/Data/List/Dedup.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Data/List/Dedup.lean:138]: I assume we don't need to wait for an answer here? Of course, if @Timeroot comes along with a shorter proof after this is merged, we can always make a golf PR.

**Review by YaelDillies — APPROVED** [review]
Looks good now

maintainer merge

**Comment by eric-wieser** [comment]
maintainer merge

The conspiracy for bots to ignore Yael continues

**Comment by mattrobball** [comment]
!bench

**Comment by mattrobball** [comment]
I would guess the import change and the single simp are good but benched so I don't have to speculate. If it comes back reasonable, please merge. Thanks!

bors delegate+

**Pull request closed without merging**